File: haskell/glue/src/Glue/Lib/List/Take.hs
module Glue.Lib.List.Take where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

take :: IR Eval
take = NativeFunc takeImpl

takeImpl :: IR Eval -> Eval (IR Eval)
takeImpl countIR = pure $ NativeFunc (takeFrom countIR)

takeFrom :: IR Eval -> IR Eval -> Eval (IR Eval)
takeFrom count list = case (count, list) of
    (Integer n, List xs) -> do
        if n < 0
            then throwError $ wrongArgumentType ["non-negative integer"]
            else pure $ List (Prelude.take (fromIntegral n) xs)
    _ -> throwError $ wrongArgumentType ["number", "list"]

File: dart/glue/lib/src/lib/list/take.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Take function - returns first N elements of a list
/// Mirrors Haskell Glue.Lib.List.Take.take exactly
Ir take = IrNativeFunc(takeImpl);

/// Take function implementation
/// Mirrors Haskell Glue.Lib.List.Take.takeImpl exactly
Eval<Ir> takeImpl(Ir countIr) {
  return Eval.pure(IrNativeFunc(takeFrom(countIr)));
}

/// Helper function for list argument
/// Mirrors Haskell Glue.Lib.List.Take.takeFrom exactly
Eval<Ir> Function(Ir) takeFrom(Ir count) {
  return (Ir list) => switch ((count, list)) {
    (IrInteger(value: final n), IrList(elements: final elements)) =>
      n < 0
          ? throwError(wrongArgumentType(['non-negative integer']))
          : Eval.pure(
              IrList(
                n >= elements.length
                    ? elements.toList()
                    : elements.take(n).toList(),
              ),
            ),
    _ => throwError(wrongArgumentType(['number', 'list'])),
  };
}
